### PR TITLE
Template: Tabular vis

### PR DESCRIPTION
**Pull request by houxinli**
## Template for Tabular Vis
- few lines of code of specification
- support rendering large tables with dynamic data fetching
- automatic styles
- not supporting predicated table yet (will be in next PR)

## Below is old content

required argument: [name, db, fields]

usage:
`p.addTable(new Table({name: "table0", db:"db0", fields:["col0", "col1", "col2"]})`

I let the Table class inherit from Layer, so technically `canvas.addLayer(new Table(args))` is doable. The only problem is that I use the `renderingParams` pass the parameters that transform function and rendering function need, and `p.addRenderingParams(table.renderingParams)`is called in p.addTable()

Haven't found a proper way to let user define a complicated query while making sure that the columns meets with the query.

I intend to bring in template default css functionality in my next pull request.

Open for suggestions, criticism and discussion

### CODE REVIEW

**Comment by tracyhenry** [comment]
thanks, I'll take a look tonight.

**Comment by houxinli** [comment]
Good idea, maybe we could open up a template-examples folder

**Comment by tracyhenry** [comment]
> Good idea, maybe we could open up a template-examples folder

yes, name it as `template-api-examples`. Be careful about the path for the config now, one more `../` is needed. Also move `nba_autodd` there. Thanks!

File: compiler/src/template-api/Table.js [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    ) {
+        throw new Error("fields and heads length not equal");
+    } else if (th_args.th_h <= 0) {
+        throw new Error("heading height must be positive");
+    } else {
+        this.heads = th_args;
+    }
+
+    var tableRenderingParams = {
+        // table_name: {
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: is bracket necessary here?

File: compiler/src/template-api/Table.js [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    this.cell_h = args.cell_h || 40;
+    this.x = args.x || 0;
+    this.y = args.y || 0;
+    this.db = args.db;
+
+    var sum_width = 0;
+    var centroid_x = 0;
+    if (typeof args.width === "number") {
+        sum_width = args.width;
+        centroid_x = this.x + args.width / 2;
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: minor: this statement to calculate `centroid_x` is repeated in each if/else branch, can be put after this if/else

File: compiler/src/template-api/Table.js [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    } else if (typeof args.width === "object") {
+        for (key in args.width) {
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: I suppose `key` here is a column name?

File: compiler/src/template-api/Table.js [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+            th_h: 0,
+            names: []
+        };
+    } else if (th_args == "auto") {
+        this.heads = {
+            th_h: this.cell_h,
+            names: args.fields
+        };
+    } else if (!("th_h" in th_args || "names" in th_args)) {
+        throw new Error("in complete heading definition");
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: minor: add "Constructing Table" in the error message

File: compiler/src/template-api/Table.js [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    } else if (typeof th_args.th_h !== "number") {
+        throw new Error("heading height must be number");
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: same: add "Constructing Table" in the error message

File: compiler/src/template-api/Table.js (hunk 2 above)
Comment (houxinli) [reply compiler/src/template-api/Table.js]: Sure

File: compiler/src/template-api/Table.js (hunk 5 above)
Comment (tracyhenry) [reply compiler/src/template-api/Table.js]: There are a bunch in the rest of the file as well. I'll skip picking them out individually.

File: compiler/src/template-api/Table.js [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    } else {
+        this.heads = th_args;
+    }
+
+    var tableRenderingParams = {
+        // table_name: {
+        [table_name]: {
+            x: this.x,
+            y: this.y,
+            heads: this.heads,
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: I'm a bit confused by `heads` here. Seems like `heads` is simply a dictionary containing both `cell_h` and `fields`?

**Comment by tracyhenry** [comment]
The rest looks pretty good to me. I'll do some testing myself tomorrow, then we can merge this.

File: compiler/src/template-api/Table.js [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+            th.append("text")
+                .text(th_Names[index])
+                .attr("dx", th_W[index] / 2)
+                .attr("y", th_H / 2)
+                .attr("dy", "0.5em");
+        }
+    }
+}
+
+function getBodyStringOfFunction(func) {
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: This function seems to be used in every template, so let's move it to `template-api/Renderers.js`

**Comment by houxinli** [comment]
> yes, name it as `template-api-examples`. Be careful about the path for the config now, one more `../` is needed.

Didn't get it... where would you want the folderto be placed? `compiler/examples/template-api-examples` or `compiler/src/template-api/template-api-examples`?

**Comment by tracyhenry** [comment]
> > yes, name it as `template-api-examples`. Be careful about the path for the config now, one more `../` is needed.
> 
> Didn't get it... where would you want the folderto be placed? `compiler/examples/template-api-examples` or `compiler/src/template-api/template-api-examples`?

As we discussed, we can use `compiler/examples/template-api-examples`, and just put js files there (no subdirectories). That way references to components (e.g. Project, Transform, etc) don't need to be changed.

File: compiler/src/template-api/Table.js (hunk 6 above)
Comment (houxinli) [reply compiler/src/template-api/Table.js]: user can rename fields here, or make table headers taller or shorter than common lines

File: compiler/src/template-api/Table.js [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+        sum_width = args.width;
+    } else if (typeof args.width === "object") {
+        if (Array.isArray(args.width)) {
+            // ES6 grammer for array
+            if (args.width.length != args.fields.length)
+                throw new Error(
+                    "Constructing Table: incompatible length between width and fields"
+                );
+            for (var index of args.width) {
+                if (typeof args.width[index] !== "number") {
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: `index` seems to be a value, rather than an index

File: compiler/src/template-api/Table.js [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+                    throw new Error(
+                        "Constructing Table: width's value must be number"
+                    );
+                key_index = args.fields.indexOf(key);
+                if (key_index >= 0) {
+                    widths[key_index] = args.width[key];
+                } else {
+                    throw new Error(
+                        "Constructing Table: width field not given in fields:",
+                        key
```
Comment (tracyhenry) [inline compiler/src/template-api/Table.js]: `key` should be concatenated rather than as a second parameter?

**Pull request merged**